index.d.ts

```diff
@@ -0,0 +1,88 @@
+// Project: https://github.com/expressjs/session
+// Definitions by: Hiroki Horiuchi <https://github.com/horiuchi>
+//                 Jacob Bogers <https://github.com/jacobbogers>
+//                 Naoto Yokoyama <https://github.com/builtinnya>
+//                 Ryan Cannon <https://github.com/ry7n>
+//                 HoldYourWaffle <https://github.com/HoldYourWaffle>
+
+import express from 'express'
+import node from 'events'
+
+
+
+declare global {
+  namespace Express {
+    interface Request { //inject additional properties on express.Request
+      session?: Session;
+      sessionID?: string;
+    }
+  }
+
+  export interface SessionData {
+    //seperate interface to allow application-specific declaration merging (see README)
+  }
+}
+
+
+
+interface SessionOptions {
+  secret: string | string[];
+
+  genid?(req: express.Request): string;
+  name?: string;
+  store?: Store;
+  cookie?: express.CookieOptions;
+  rolling?: boolean;
+  resave?: boolean;
+  proxy?: boolean;
+  saveUninitialized?: boolean;
+  unset?: string;
+}
+
+export default function session(options?: SessionOptions): express.RequestHandler;
+
+
+
+interface Session extends SessionData {
+  id: string;
+  cookie: SessionCookie;
+
+  regenerate(callback: (err: any) => void): Session;
+  destroy(callback: (err: any) => void): Session;
+  reload(callback: (err: any) => void): Session;
+  save(callback?: (err: any) => void): Session;
+  touch(): Session;
+}
+
+interface SessionCookie extends express.CookieOptions { //XXX actually a class
+  originalMaxAge: number;
+}
+
+
+
+export abstract class Store extends node.EventEmitter {
+  constructor(config?: any);
+
+  regenerate: (req: express.Request, callback: (err?: any) => any) => void;
+  load: (sid: string, callback: (err: any, session?: SessionData) => any) => void;
+  createSession: (req: express.Request, sess: SessionData) => void;
+
+  abstract get: (sid: string, callback: (err: any, session?: SessionData | null) => void) => void;
+  abstract set: (sid: string, session: SessionData, callback?: (err?: any) => void) => void;
+  abstract destroy: (sid: string, callback?: (err?: any) => void) => void;
+  abstract all: (callback: (err: any, obj?: { [sid: string]: SessionData; } | null) => void) => void;
+  abstract length: (callback: (err: any, length?: number | null) => void) => void;
+  abstract clear: (callback?: (err?: any) => void) => void;
+  abstract touch: (sid: string, session: SessionData, callback?: (err?: any) => void) => void;
+}
+
+export class MemoryStore extends Store {
+  get: (sid: string, callback: (err: any, session?: SessionData | null) => void) => void;
+  set: (sid: string, session: SessionData, callback?: (err?: any) => void) => void;
+  destroy: (sid: string, callback?: (err?: any) => void) => void;
+  all: (callback: (err: any, obj?: { [sid: string]: SessionData; } | null) => void) => void;
+  length: (callback: (err: any, length?: number | null) => void) => void;
```

Comment:
So the `length` will never be `null` for `MemoryStore`. Impl here: https://github.com/expressjs/session/blob/master/session/memory.js#L134 Not sure if that is what this definition is implying or not.

Reply:
`length?: number | null` means that there's an *optional* parameter `length` (could be `undefined`, see my earlier comment) that can be of type `number` **or** `null` (`|` defines a [union type](https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types), effectively an `or`). If you didn't understand any of that: doesn't matter, it's probably wrong in this case.

I didn't change this type from the DT one, but now that you mention it it seems a little backwards. Would a regular (non-optional) `length` parameter of type `number` make more sense?

Reply:
So right now I'm just looking through and trying to eye ball this. I have never used typescript and don't understand it's specifics, so cannot answer anything related to it. Maybe if there were instructions for how I could use this in an example app I could play around? Not sure if that would help me answer your question or not.

Reply:
I think the [Typescript in 5 minutes](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html) and [Basic Types](https://www.typescriptlang.org/docs/handbook/basic-types.html) pages from the official documentation would be your best bet for setting up and understanding a basic TypeScript project. Keep in mind that you should probably set `strictNullChecks` to `true` in your `tsconfig.json` to get the best results.

To use these new definitions in your own project you could run `npm install github:HoldYourWaffle/express-session#integrate-update-typings` to install from my fork directly (although you'd have to re-run this every time I create a new commit, since npm doesn't properly support updates for git repositories yet), or you could just manually copy the [`index.d.ts`](https://github.com/HoldYourWaffle/express-session/blob/integrate-update-typings/index.d.ts) file from my fork into your own project (not sure if this will work immediately out of the box, feel free to ask for help if you can't get it to work).

Will all of this help you with answering my question? Probably not, since it's for one a simple typing choice and secondly it's a semantic choice: is it valid for implementations to pass in certain types?

To rephrase my question: is it valid for implementations to pass `null` and/or `undefined` as `length` to the callback?